src/compiler/checker.ts

```diff
@@ -28921,7 +28918,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         while (
             flowContainer !== declarationContainer && (flowContainer.kind === SyntaxKind.FunctionExpression ||
                 flowContainer.kind === SyntaxKind.ArrowFunction || isObjectLiteralOrClassExpressionMethodOrAccessor(flowContainer)) &&
-            (isConstantVariable(localOrExportSymbol) && type !== autoArrayType || isParameter && !isSymbolAssigned(localOrExportSymbol))
+            (isConstantVariable(localOrExportSymbol) && type !== autoArrayType || isParameter && !isSomeSymbolAssigned(rootDeclaration))

```

Comment:
It looks like the change to `isSymbolAssigned` behavior, to aggregate `isAssigned` at the root declaration level, might not be the correct thing to do here, for this particular usage of `isSymbolAssigned`, as evidenced by this breaking change: https://github.com/microsoft/TypeScript/pull/56313#issuecomment-1841559785.

Maybe we want to keep marking symbols with `isAssigned`, and also aggregate that info at the root declaration level, since `getNarrowedTypeOfSymbol` needs the aggregate info but `checkIdentifier` needs the symbol info.

Here's a more minimal repro of the break linked above:
```ts
function ff({ a, b }: { a: string | undefined, b: () => void }) {
  if (a !== undefined) {
    b = () => {
      const x: string = a;
    }
  }
}
```

Reply:
Heh, I really need to start reading through all of the existing comments at once instead of reading through them one by one. It would save me a few minutes here 😅